optuna/_hypervolume/wfg.py

```diff
@@ -2,61 +2,75 @@
-    ) -> float:
-        if not np.all(np.isfinite(reference_point)):
-            # reference_point does not have nan, because BaseHypervolume._validate will filter out.
-            return float("inf")
+    This class exactly calculates the hypervolume for any dimension.
+    For 3 dimensions or higher, the WFG algorithm will be used.
+    Please refer to ``A fast way of calculating exact hypervolumes`` for the WFG algorithm.
 
-        if not assume_pareto:
-            unique_lexsorted_sols = np.unique(solution_set, axis=0)
-            sorted_pareto_sols = unique_lexsorted_sols[_is_pareto_front(unique_lexsorted_sols)]
-        else:
-            sorted_pareto_sols = solution_set[solution_set[:, 0].argsort()]
+    .. note::
+        This class is used for computing the hypervolumes of points in multi-objective space.
+        Each coordinate of each point represents a ``values`` of the multi-objective function.
 
-        self._reference_point = reference_point.astype(np.float64)
-        if self._reference_point.shape[0] == 2:
-            return _compute_2d(sorted_pareto_sols, self._reference_point)
+    .. note::
+        We check that each objective is to be minimized. Transform objective values that are
+        to be maximized before calling this class's ``compute`` method.
 
-        return self._compute_hv(sorted_pareto_sols)
+    """
 
-    def _compute_hv(self, sorted_sols: np.ndarray) -> float:
-        assert self._reference_point is not None
-        inclusive_hvs = np.prod(self._reference_point - sorted_sols, axis=-1)
+    if not np.all(solution_set <= reference_point):
+        raise ValueError(
+            "All points must dominate or equal the reference point. "
+            "That is, for all points in the solution_set and the coordinate `i`, "
+            "`solution_set[i] <= reference_point[i]`."
+        )
+    if not np.all(np.isfinite(reference_point)):
+        # reference_point does not have nan, because BaseHypervolume._validate will filter out.
+        return float("inf")
+
+    def _compute_hv(sorted_sols: np.ndarray) -> float:
+        inclusive_hvs = np.prod(reference_point - sorted_sols, axis=-1)
         if inclusive_hvs.shape[0] == 1:
             return float(inclusive_hvs[0])
         elif inclusive_hvs.shape[0] == 2:
             # S(A v B) = S(A) + S(B) - S(A ^ B).
-            intersec = np.prod(self._reference_point - np.maximum(sorted_sols[0], sorted_sols[1]))
+            intersec = np.prod(reference_point - np.maximum(sorted_sols[0], sorted_sols[1]))
             return np.sum(inclusive_hvs) - intersec
 
         limited_sols_array = np.maximum(sorted_sols[:, np.newaxis], sorted_sols)
         return sum(
-            self._compute_exclusive_hv(limited_sols_array[i, i + 1 :], inclusive_hv)
+            _compute_exclusive_hv(limited_sols_array[i, i + 1 :], inclusive_hv)
             for i, inclusive_hv in enumerate(inclusive_hvs)
         )
```

Comment:
How about moving this function (and others placed below) outside of `compute_hypervolume`?

Reply:
@HideakiImamura 
I addressed your suggestion!

https://github.com/optuna/optuna/pull/5504/commits/718b56e979aa7638d17a90ec67804350d8f220fd

@eukaryo 
As I transferred the nested functions outside, we do not need the followup anymore:)